controllers/cloudstackfailuredomain_controller.go

```diff
@@ -130,21 +140,222 @@ func (r *CloudStackFailureDomainReconciliationRunner) ClearMachines() (ctrl.Resu
 	if err := r.K8sClient.List(r.RequestCtx, machines, client.MatchingLabels{infrav1.FailureDomainLabelName: r.ReconciliationSubject.Name}); err != nil {
 		return ctrl.Result{}, err
 	}
-	// Deleted CAPI machines for CloudStack machines found.
-	for _, machine := range machines.Items {
+	r.Log.Info(fmt.Sprintf("Clear machine: %d found.", len(machines.Items)))
+	if len(machines.Items) > 0 {
+		if result, err := checkClusterReady(r); err != nil {
+			return result, err
+		}
+	}
+
+	result, err := triggerEtcdClusterRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	result, err = triggerControlPlaneRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	result, err = triggerMachineDeploymentRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	if len(machines.Items) > 0 {
+		return r.RequeueWithMessage("FailureDomain still has machine(s) in it.")
+	}
+	return ctrl.Result{}, nil
+}
+
+func triggerEtcdClusterRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	var etcdMachine infrav1.CloudStackMachine
+	etcdMachineFound := false
+	for _, machine := range machines {
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == "EtcdadmCluster" {
+				etcdMachineFound = true
+				etcdMachine = machine
+			}
+		}
+	}
+	if !etcdMachineFound {
+		r.Log.Info("Clear machine: no etcd VM found.")
+		return ctrl.Result{}, nil
+	}
+	for _, ref := range etcdMachine.OwnerReferences {
+		if ref.Kind == "EtcdadmCluster" {
+			r.Log.Info("Clear machine: getting etcdadmcluster ...")
+			etcdadmCluster := &unstructured.Unstructured{}
+			etcdadmCluster.SetGroupVersionKind(schema.FromAPIVersionAndKind(ref.APIVersion, ref.Kind))
+			if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: etcdMachine.Namespace, Name: ref.Name}, etcdadmCluster); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info("Clear machine: retrieved etcdadmcluster.")
+			csMachineTemplateName, ok := etcdMachine.Annotations["cluster.x-k8s.io/cloned-from-name"]
+			if !ok {
+				result, err := r.RequeueWithMessage("annotations['cluster.x-k8s.io/cloned-from-name'] not found.", "cloudstackmachine", etcdMachine.Name)
+				return result, err
+			}
+			r.Log.Info(fmt.Sprintf("Clear machine: machine template name %s decided", csMachineTemplateName))
+			csMachineTemplateNameInEtcdadmCluster, found, err := unstructured.NestedString(etcdadmCluster.Object, "spec", "infrastructureTemplate", "name")
+			if err != nil || !found {
+				return ctrl.Result{}, errors.New("etcdadmcluster spec.infrastructureTemplate.name not found or not string")
```

Comment:
maybe also log the etcdadmcluster name for debugging ease? It might also help to return the original error for context